File: backend/app/models/schedule.py
from datetime import datetime

from sqlalchemy import Boolean, DateTime, ForeignKey, Integer, JSON, String, Text
from sqlalchemy.orm import Mapped, mapped_column

from app.models.base import Base, TimestampMixin


class SchedulePlan(TimestampMixin, Base):
    __tablename__ = "schedule_plans"
    name: Mapped[str] = mapped_column(String(500), nullable=False)
    status: Mapped[str] = mapped_column(String(20), nullable=False, default="active")
    is_recurring: Mapped[bool] = mapped_column(Boolean, default=False)
    recurrence_cron: Mapped[str | None] = mapped_column(String(100), nullable=True)
    recurrence_count: Mapped[int] = mapped_column(Integer, default=0)
    next_run_at: Mapped[datetime | None] = mapped_column(DateTime, nullable=True)


class ScheduleTask(TimestampMixin, Base):
    __tablename__ = "schedule_tasks"
    plan_id: Mapped[str] = mapped_column(
        String(36), ForeignKey("schedule_plans.id"), nullable=False
    )
    orchestration_id: Mapped[str] = mapped_column(String(36), nullable=False)
    agent_id: Mapped[str | None] = mapped_column(String(36), nullable=True)
    wagent_id: Mapped[str | None] = mapped_column(String(36), nullable=True)
    wagent_version: Mapped[int | None] = mapped_column(Integer, nullable=True)
    status: Mapped[str] = mapped_column(String(20), nullable=False, default="pending")
    priority: Mapped[str] = mapped_column(String(10), default="medium")
    scheduled_at: Mapped[datetime] = mapped_column(DateTime, nullable=False)
    started_at: Mapped[datetime | None] = mapped_column(DateTime, nullable=True)
    completed_at: Mapped[datetime | None] = mapped_column(DateTime, nullable=True)
    input_params: Mapped[dict | None] = mapped_column(JSON, nullable=True)
    output_result: Mapped[dict | None] = mapped_column(JSON, nullable=True)
    error_message: Mapped[str | None] = mapped_column(Text, nullable=True)
    retry_count: Mapped[int] = mapped_column(Integer, default=0)
    max_retries: Mapped[int] = mapped_column(Integer, default=3)
    dependencies: Mapped[dict] = mapped_column(JSON, default=list)
    confirm_deadline: Mapped[datetime | None] = mapped_column(DateTime, nullable=True)
    confirm_action: Mapped[str | None] = mapped_column(String(20), nullable=True)
    execution_log: Mapped[str | None] = mapped_column(Text, nullable=True)

File: frontend/src/types/schedule.ts
export interface SchedulePlan {
  id: string;
  name: string;
  status: string;
  is_recurring: boolean;
  recurrence_cron?: string;
  recurrence_count: number;
  next_run_at?: string;
  created_at: string;
  updated_at: string;
}

export interface ScheduleTask {
  id: string;
  plan_id: string;
  plan_name?: string;
  orchestration_id: string;
  agent_id?: string;
  wagent_id?: string;
  agent_name?: string;
  wagent_version?: number;
  status: string;
  priority: string;
  scheduled_at: string;
  started_at?: string;
  completed_at?: string;
  input_params?: Record<string, unknown>;
  output_result?: unknown;
  error_message?: string;
  retry_count: number;
  max_retries: number;
  dependencies: string[];
  confirm_deadline?: string;
  confirm_action?: string;
  execution_log?: string;
  created_at: string;
  updated_at: string;
}
